### DISCUSSION TITLE
Distributed framework

### DISCUSSION THREAD

**Discussion by bmalezieux** [deepinv/deepinv#905] [Ideas]
# Distributed Framework for DeepInverse: Design Discussion

Hi everyone,

We've been working on a distributed computing framework for DeepInverse and would like to discuss the design to ensure we build something that integrates well with the library's philosophy and meets users' needs.

### What's Distributed Right Now

The current implementation provides distributed versions of the core components needed for inverse problem solving:

1. **DistributedPhysics** / **DistributedLinearPhysics**
   - Distributes multiple physics operators across processes
   - Each process owns a subset of operators (e.g., Process 0 handles operators 0,2,4; Process 1 handles 1,3,5)
   - Supports both local (fast) and global (compatible) forward operations

2. **DistributedMeasurements**
   - Stores only local measurement shards on each process
   - Aligned with the distributed physics operators

3. **DistributedSignal**
   - Represents the reconstruction signal (x) that's replicated across all processes
   - Automatically synchronized after updates via all-reduce/broadcast
   - Provides a consistent view of the signal across all processes

4. **DistributedDataFidelity**
   - Computes data fidelity terms using only local physics/measurements
   - Loss: local computations + single all-reduce for global sum
   - Gradient: local VJP computations + single all-reduce for global gradient

5. **DistributedPrior**
   - Uses pluggable signal processing strategies to split work
   - Currently supports spatial tiling strategies (2D/3D)
   - Each process works on different patches of the image
   - Handles overlapping patches with averaging for smooth reconstruction

### How It Works

The framework uses a **hybrid distribution strategy**:

- **Physics operators**: distributed by operator index (different processes handle different measurement operators)
- **Signal (x)**: replicated across all processes with automatic synchronization
- **Priors**: distributed by spatial patches (different processes handle different regions of the image)

**Communication pattern:**
- Data fidelity operations: one all-reduce per iteration (for gradient)
- Prior operations: multiple all-reduce operations for patch assembly and reduction
- Signal updates: automatic broadcast/all-reduce after modifications

### Integration with PnP Framework

Here's how it integrates into a typical Plug-and-Play reconstruction workflow:

```python
# 1. Initialize distributed context
with DistributedContext(seed=42) as ctx:
    
    # 2. Create physics operators and measurements (can use StackedPhysics)
    stacked_physics, measurements, clean_image = create_physics_and_measurements(
        ctx.device, img_size=(512, 512)
    )
    
    # 3. Define your prior (standard DeepInverse prior)
    denoiser = DRUNet(pretrained="download").to(ctx.device)
    pnp_prior = PnP(denoiser=denoiser)
    
    # 4. Configure distributed components via factory API
    factory_config = FactoryConfig(
        physics=stacked_physics,      # Automatically extracts operators
        measurements=measurements,      # TensorList or list of tensors
        data_fidelity=L2(),            # Standard data fidelity
    )
    
    tiling_config = TilingConfig(
        patch_size=128,
        receptive_field_size=32,
    )
    
    # 5. Build distributed bundle (one call creates everything)
    distributed_bundle = make_distrib_bundle(
        ctx,
        factory_config=factory_config,
        signal_shape=(B, C, H, W),
        prior=pnp_prior,
        tiling=tiling_config,
    )
    
    # 6. Run PnP iterations using distributed components
    for it in range(num_iterations):
        # Data fidelity gradient (distributed)
        grad = distributed_bundle.data_fidelity.grad(distributed_bundle.signal)
        
        # Gradient descent
        new_data = distributed_bundle.signal.data - lr * grad
        distributed_bundle.signal.update_(new_data)
        
        # Denoising with distributed prior
        denoised = distributed_bundle.prior.prox(
            distributed_bundle.signal, 
            sigma_denoiser=denoiser_sigma
        )
        distributed_bundle.signal.update_(denoised)
```

## Questions

We'd like to discuss what the **minimal atomic features** should be and how to design the **right API** for users.

### 4. Should We Support Partial Distribution?

Currently, you either distribute everything (physics + prior) or nothing. Should we support:
- Distributing only physics operators (signal replicated, prior runs on full signal)
- Distributing only priors (physics on rank 0, prior distributed spatially)
- Mixed strategies (some operators distributed, others replicated)

### 5. How Should This Integrate with Existing Workflows?

--------

**Use Cases to Consider:**

Here are some use cases that should inform the API design:

1. **Large-scale radio interferometry**: 1000s of visibility operators, GB-scale measurements
2. **Medical imaging**: 3D CT/MRI with volumetric priors, memory constraints per GPU
3. **Research**: Easy to experiment with different distribution strategies

**Current Repository Structure:**
```
deepinv/distrib/
├── __init__.py                    # Public API
├── distrib_framework.py           # Core distributed classes
├── utils.py                       # Factory API and configs
└── distribution_strategies/
    ├── strategies.py              # Signal processing strategies
    └── utils.py                   # Tiling utilities
```

**Example Usage:**
See `examples/distrib/demo_pnp_distributed.py` for a complete working example.

**Comment by tachella** [comment]
Hey @bmalezieux, many thanks for this summary!

Here is a summary of the use cases that I have in mind:


### 0. Distributed Physics
Some physics might be too heavy to compute on a single device. We could distribute the physics across devices as

```
stacked_physics = dinv.physics.stack([physics1, physics2, ...])

distributed_physics = dinv.distrib.distribute(stacked_physics, ...)

x = torch.ones(1, 3, 4096, 4096, device='cuda')

# applies physics in parallel on different GPUs, and assembles back 
y = physics(x)
# y can be either on a single device or distributed (according to how big it is)

# applies A^tA in parallel on different GPUs, and sums 
x2 = physics.A_adjoint_A(x)

```

### 1. Distributed Denoiser

The user wants to run a `deepinv.models.Denoiser` on a very large image/volume in parallel with multiple GPUs. 
I imagine something like

```
denoiser = dinv.models.DRUNet()

tiling_config = dinv.distrib.TilingConfig(...)
distributed_denoiser = dinv.distrib.distribute(denoiser, tiling_config, ...)

y = torch.ones(1, 3, 4096, 4096, device='cuda')

# this tiles the image, applies denoiser in parallel on different GPUs, and assembles back 
x_hat = distributed_denoiser(y, sigma)

# plot result
dinv.utils.plot(x_hat)

```
Are we forced to convert y to a distributed signal here? Or can this logic be handled inside the distributed denoiser?

### 2. Distributed optim with explicit for-loop

This is quite similar to what you have already done, combining both points before

```
prior = dinv.optim.PnP(distributed_denoiser) # should also work for distributed_prior

x = torch.ones(1, 3, 4096, 4096, device='cuda')

for i in range(100):
      x = x - eta * data_fidelity.grad(x, y, distributed_physics)
      x = prior.prox(x)

# plot result
dinv.utils.plot(x_hat)

```
This setting doesn't cover non-PnP priors..

### 2. Distributed optim with black-box models

The library allows defining 'standard' optimization methods in one line of code 

```
model = dinv.optim.optim_builder(
    iteration="PGD",
    prior=prior,
    data_fidelity=data_fidelity,
    max_iter=max_iter
)

model = dinv.distrib.distribute(model, tiling_config, ...)

# this computes the optim algorithm by distributing prior and data_fid
x_hat = model(y, distributed_physics)
```
@samuro95 is working on a refactoring of the `optim_builder` (see #592), such that we write directly `model = dinv.optim.PGD(...)`. The same should be done for unfolded networks (in #592, this should amount to simply adding a flag, `unfold=True`).


### 3. Distributed Reconstructor

Taking a final step in generalization, we could also distribute general `dinv.models.Reconstructor`s (not only optimization-based models, such as PnP and unfolded)

```
model = dinv.models.ArtifactRemoval(dinv.models.DRUNet())
# or
model = dinv.models.RAM()

model = dinv.distrib.distribute(model, tiling_config, ...)
# this computes the model in a distributed way (to discuss)
x_hat = model(y, distributed_physics)
```

**Reply by bmalezieux** [reply]
> Are we forced to convert y to a distributed signal here? Or can this logic be handled inside the distributed denoiser?

For sure with one pass in the denoiser we don't need it, but then for several passes (for instance in a PnP) we need to make sure that the signal is consistent across ranks. I'll think about a way to hide this.

**Comment by romainvo** [comment]
I thought I might continue the discussion with respect to my comment in the PR (https://github.com/deepinv/deepinv/pull/790#discussion_r2538850292) here
> I get the need for a `DistributedDataFidelity` now. Here this step is less efficient than in your previous commit [1996382](https://github.com/deepinv/deepinv/commit/199638210d9dcd4a72ba55d2883d5b7eeb6d712f), ie it does two reduce/gather operations instead of only one ?
> 
> A solution I'm thinking if we absolutely do not want to write a `DistributedDataFidelity` strategy, is to dynamically control the behaviour of `A_vjp` and `A` given the scope it is called from, ie if it is called by `DataFidelity.grad` we switch to `_local` computations and defer reducing to after `A_vjp`. With [inspect](https://docs.python.org/3/library/nspect.html) and some decorator it could be feasible, but might be ugly.

With the current version of the framework, we _implicitly_ force the user to gather/reduce for every call to `A` or `A_adjoint`, which is less efficient that timing it properly. 

The example I have in mind is when calling `data_fidelity.grad(x, measurements, distributed_physics)`:
```python

# does grad = physics.A_vjp(x, self.d.grad(physics.A(x), y))
grad = data_fidelity.grad(x, measurements, distributed_physics)

# What it does
# 1. call physics.A_local on each rank
# 2. all_gather local results to build global Ax
# 3. compute global grad (on every rank)
# 4. calls A_vjp on global grad
# 5. distribute A_vjp(grad) across ranks
# 6. all_reduce local results
```
here step 2 is useless and could be skipped. The grad should be computed per rank (step 3.), as well as the call to `A_vjp` (step 4.). Then we call an all_reduce operation.

Another point is if the user, for whatever reason, wishes to compute some statistics on the sub-problems. By forcing gather/reduce we remove this option. 
- There is always the option to call the `_local` counterparts of `A` and `A_adjoint`. However, the problem is that most of `deepinv` functionalities rely on calls to `A` and `A_adjoint` explicitly, which makes it difficult to use the implemented tools (solvers or whatever) on sub-problems. 

A solution, would be to remove the `_local` functions, and make `A`, `A_adjoint` and co' local by default, which is closer to what torch [DistributedDataParallell](https://docs.pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) actually does, i.e. a distributed `nn.Module.forward` method does not reduce local results automatically.
- to make it easy to use for the non-expert, we could give the option when building a `DistributedPhysics` to gather/reduce for every call to `A` and `A_adjoint`.
- we could implement special behaviour, which times communications efficiently, depending on the parent scope of the calls, e.g. in case it is called by a `.grad` method.
- for the user who wants to tinker, we can give the option to activate/deactivate implicit communication calls on demand.

@bmalezieux @tachella what do you think ? Please tell me @bmalezieux  if I missed something, or misunderstood something in the proces.

**Reply by bmalezieux** [reply]
Yes, without a distributed data fidelity, we can't keep the grad and vjp computations local before gather/reduce, and it will impact the performance. I think it's fine to have a distributed data fidelity and rewrite the main components (fn and grad) such that we only gather/reduce when we have the local gradients and not before. 

Regarding the automatic reduce, maybe we could reduce by default but give the option to keep computations local as you said. This way it's still easy for non-expert users and others will be able to tune their algorithms.
